### PR TITLE
Naf obj #122

### PR DESCRIPTION
**Pull request by namirfa**
Fixes Issue #122 .

Adds the ability to retrieve the input frame size from the .meta file. If no meta file is provided, then the standard 416 default value is used. For backwards compatibility, the user can still override both the default and the meta file's input frame size by editing the edu.ml.tensorflow.Config.FRAME_SIZE value.

I was unable to write automated tests that check the behavior when the default value is overridden. I was also unable to add tests that demonstrate the behavior if the meta file has invalid height/width values. These were both manually tested and worked as expected.

### CODE REVIEW

**Comment by namirfa** [comment]
Additionally, I did not find any other obvious values that we can pull from the .meta file. I will double check now to see if there is anything else worth adding here.

File: objectRecognitionSource/src/main/java/edu/ml/tensorflow/ObjectDetector.java [hunk 1]
```diff
@@ -45,6 +42,11 @@
     // "year-month-date-hour-minute-seconds"
     private static SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd--HH-mm-ss");
     
+    // Getting meta config options for YOLO Processor
+    public MetaBasedConfig metaConfigOptions = new MetaBasedConfig();
+    private final float MEAN = metaConfigOptions.mean;
```
[the remark this answers is not in the document]
Comment (namirfa) [reply objectRecognitionSource/src/main/java/edu/ml/tensorflow/ObjectDetector.java]: Just getting a chance to read this. So instead here, I should just leave the original MEAN as a constant taken from the Config file (how it was before), and not set it as part of the MetaBasedConfig.

Let me know if that seems about right.

[the remark this answers is not in the document]
Comment (fhcarter) [reply objectRecognitionSource/src/main/java/edu/ml/tensorflow/ObjectDetector.java]: @namirfa Yes, I think so.  I don't see any MEAN/mean/Mean/etc/ in any of the `.meta` or `.cfg` files, so I don't think we're gathering it from anywhere.  It's just some number that's meaningful.  Somehow?

I also looked for variations on `scale` and they don't exist either.  From some googling, I found some hints [here](https://github.com/thtrieu/darkflow/issues/38).  They appear to say that the 255 is used for normalizing the images -- I suspect that it has to do with the scaling of colors or B&W it's using.  The hint (link above) referred to that scaling number being in the range (0, 1) when running on the android version of things, so I that the 255 has to do with the number of shades involved or or something else referring to greyscale issues or something (but that's really more speculation).  It seems to be structural part of YOLO v2 (at least) that we probably shouldn't change.  It might make sense to change the name to NORMALIZATION_SCALE or something so we don't have to re-derive it each time :-).  Or maybe just add a comment to the Config file that says what it's for.  

If you read thru the graph builder & use thereof in `ObjectDetector.createNormalizerGraph(), it's decoding the JPEG into some matrix (I think), and then dividing that whole matrix (or, really, TensorFlow `Output` object) by the 255/Mean/Scale value.  So, I think it's taking the values and normalizing them into an 8-bit color/B&W scale.  I don't really think we want it to be adjustable -- and if someone wants to, they can change the Config.java file.  

So, bottom line.  Maybe I've convinced myself that it'll work as is -- pass it straight through from Config -- Don't try & read it from `.meta` (which I don't think you're doing).  Whether it has to go via `MetaBasedConfig` isn't clear.  Probably so to keep it clean, but please add a comment that says that this is not read from the `.meta` file.

File: objectRecognitionSource/src/main/java/edu/ml/tensorflow/classifier/YOLOClassifier.java [hunk 2]
```diff
@@ -46,7 +46,8 @@ public static YOLOClassifier getInstance(float thresh, double[] anchorArray, int
                 anchors = anchorArray;
             }
         }
-
+        System.out.println("Grid Size: " + gridSize);
+        System.out.println("Frame Size: " + frameSize);
```
Comment (fhcarter) [inline objectRecognitionSource/src/main/java/edu/ml/tensorflow/classifier/YOLOClassifier.java]: change to log.debug or remove, please

File: objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestYoloProcessor.java [hunk 3]
```diff
@@ -240,31 +243,31 @@ public void testMetaConfig() {
         assert ypImageSaver.objectDetector.metaConfigOptions.useMetaIfAvailable == true;
         
         // Frame Size should be 416 since this is what is included in the meta file
```
Comment (fhcarter) [inline objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestYoloProcessor.java]: comment should be 608, no?

File: objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestYoloProcessor.java [hunk 4]
```diff
@@ -948,18 +978,32 @@ public void testQuery() throws ImageProcessingException, InterruptedException {
...
+                    + "             \"metaFile\": \"" + META_FILE_608 + "\", "
+                    + "             \"type\": \"yolo\""
+                    + "         }"
+                    + "}";
 
-    List<Map> getExpectedResults() throws JsonParseException, JsonMappingException, IOException {
+    List<Map> getExpectedResults(boolean useDefault) throws JsonParseException, JsonMappingException, IOException {
         ObjectMapper m = new ObjectMapper();
-        return m.readValue(imageResultsAsString, List.class);
+        if (useDefault) {
```
Comment (fhcarter) [inline objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestYoloProcessor.java]: This will be cleaner & more extensible if, instead of passing in a boolean, just pass in string to test.  That is,
```
List<Map> getExpectedResults(String expectedResults)
...
return m.readValue(expectedResults, List.class)
```

As soon as we have to test another image or another result, the boolean -- limit of two choices -- will be insufficient.

This can be done in a subsequent PR, but better to think a bit more future-proof.

File: objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestYoloProcessor.java [hunk 5]
```diff
@@ -948,18 +978,32 @@ public void testQuery() throws ImageProcessingException, InterruptedException {
...
+            return m.readValue(imageResultsAsString, List.class);
+        } else {
+            return m.readValue(imageResultsAsString608, List.class);
+        }
     }
 
-    void verifyProcessing(YoloProcessor ypImageSaver) throws ImageProcessingException {
+    void verifyProcessing(YoloProcessor ypImageSaver, boolean useDefault) throws ImageProcessingException {
```
Comment (fhcarter) [inline objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestYoloProcessor.java]: `String expectedResults` here, too

File: objectRecognitionSource/src/main/java/edu/ml/tensorflow/ObjectDetector.java [hunk 6]
```diff
@@ -317,8 +328,8 @@ private Graph createNormalizerGraph() {
         	map.put("label", recognition.getTitle());
         	map.put("confidence", recognition.getConfidence());
         	
-        	float scaleX = (float) buffImage.getWidth() / (float) FRAME_SIZE;
-            float scaleY = (float) buffImage.getHeight() / (float) FRAME_SIZE;
+        	float scaleX = (float) buffImage.getWidth() / (float) frameSize;
+            float scaleY = (float) buffImage.getHeight() / (float) frameSize;
```
Comment (fhcarter) [inline objectRecognitionSource/src/main/java/edu/ml/tensorflow/ObjectDetector.java]: spacing/indent is off on 2nd line.  (Was before, too, but might as well fix)

File: objectRecognitionSource/src/main/java/edu/ml/tensorflow/classifier/YOLOClassifier.java (hunk 2 above)
Comment (namirfa) [reply objectRecognitionSource/src/main/java/edu/ml/tensorflow/classifier/YOLOClassifier.java]: My mistake, this was for my personal debugging. Removing these.

File: objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestYoloProcessor.java (hunk 3 above)
Comment (namirfa) [reply objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestYoloProcessor.java]: Correct, I'll change this.

File: objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestYoloProcessor.java (hunk 4 above)
Comment (namirfa) [reply objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestYoloProcessor.java]: Agreed this is a better model, I'll change it.
